service/dap/server.go

```diff
@@ -765,11 +820,14 @@ func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
 			}
 		}
 
-		s.log.Debugf("building binary at %s", debugbinary)
 		var cmd string
 		var out []byte
+		// Log debug binary build
+		if !((mode == "replay") && (s.config.Debugger.Backend == "core")) {
+			s.log.Debugf("building binary '%s' from '%s' with flags '%v'", debugbinary, program, buildFlags)
+		}
 		switch mode {
-		case "debug":
+		case "debug", "replay", "record":
```

Comment:
This is getting complicated. For rr replay, `dlv replay` command requires only the trace directory as an input. So, this is not necessary to build a binary here. For core replay, I assume the user provides the executable used when generating the core. So, build the target is again unnecessary. In either case, rebuilding a binary for replay doesn't make sense to me.

How about having 'replay' mode to have the executable path attribute? For 'core', the executable path must be provided. For 'rr', the executable path is ignored. Or, alternatively, we can have separate 'replay' and 'core' modes (consistent with dlv's commands) - to avoid confusion and mistake.

For "record" mode, the current approach doesn't allow users to use 'record' to use with 'test' or 'exec'. How about exposing rr by introducing `"backend": "rr"` as @aarzilli proposed? 

If backend is rr and the user didn't specify the trace directory, that's an error.

Reply:
Per [this thread](https://github.com/go-delve/delve/pull/2367#discussion_r590929716) I had agreed with @polinasok to remove the backend parameter altogether from the [vs code](https://github.com/golang/vscode-go/pull/89) launch options (which as of master it currently exists) and leave just the configuration options for the sake of simplicity but I see where the tradeoff goes as your points are all valid

I think we could use this thread to reach consensus on how to define this contract in a way and I can start working on that definition for a final draft of the implementation

Reply:
I had a chat with @polinasok. IIUC, in the thread @polinasok commented specifically about `"backend": "coredump"` because `"coredump"` isn't a valid dlv backend. 

I think dlv and vscode launch configs align better if approximately:

   mode -> maps to dlv subcommands; debug, exec, test, (core), (replay)
   backend -> maps to dlv `--backend`; default, native, lldb, rr

And, personally, I think the review would be fast if we split this PR to smaller scopes 

* this PR: focus on making the existing debug/exec/test mode to work with the new `backend : rr` .
    -> we need testing.
    
* next PR: focus on `mode: replay` for completeness of the `replay` feature.

* next PR: focus on `mode: core`  for `coredump` feature.

What do you think, @polinasok and @lggomez ?

Reply:
So, based on this feedback I did an initial refactor of the logic and decoupled the mixed logic into corresponding blocks for "core" and "replay" modes. As for the possibility of extracting one of these features into a separate PR (be it core dumps or record) I'll let you folks decide

Reply:
Can you please update the PR description & title to clarify the scope of this PR? Thanks @lggomez!

Reply:
Done @hyangah